Tests/MongoSwiftSyncTests/SpecTestRunner/SpecTest.swift

```diff
@@ -257,37 +328,78 @@ extension SpecTest {
 
         print("Executing test: \(self.description)")
 
-        let clientOptions = self.clientOptions ?? ClientOptions(retryReads: true)
+        let clientOptions = self.clientOptions?.toClientOptions()
 
-        if let failPoint = self.failPoint {
-            try parent.activateFailPoint(failPoint)
+        var singleMongos = true
+        if let useMultipleMongoses = self.useMultipleMongoses, useMultipleMongoses == true {
+            singleMongos = false
         }
-        defer { parent.disableActiveFailPoint() }
 
-        let client = try MongoClient.makeTestClient(options: clientOptions)
+        let client = try MongoClient.makeTestClient(
+            MongoSwiftTestCase.getConnectionString(singleMongos: singleMongos), options: clientOptions
+        )
         let monitor = client.addCommandMonitor()
 
-        let db = client.db(dbName)
-        var collection: MongoCollection<Document>?
-
         if let collName = collName {
-            collection = db.collection(collName)
+            _ = try? client.db(dbName).createCollection(collName)
+            // Run the distinct command before every test to prevent `StableDbVersion` error in sharded cluster
+            // transactions. This workaround can be removed once SERVER-39704 is resolved.
+            _ = try? client.db(dbName).collection(collName).distinct(fieldName: "_id")
+        }
+
+        if let failPoint = self.failPoint {
+            try parent.activateFailPoint(failPoint)
+        }
+        defer { parent.disableActiveFailPoint() }
+
+        var sessions = [String: ClientSession]()
+        for session in Self.sessionNames {
+            sessions[session] = client.startSession(options: self.sessionOptions?[session])
         }
 
+        var sessionIds = [String: Document]()
```

Comment:
given the way this map is actually used, I think it should probably be flipped to a `[Document: String]`, since we're using the ids to look up session names.

Reply:
Done. One thing to note is that I've had to nil-coalesce the `id` since it's optional in ClientSession. This default value should never occur, however, since we explicitly start the sessions above.

Reply:
So it actually can occur because `startSession` doesn't actually start the session; that only happens upon operation execution (see the discussion in slack about this). So if the `validateExecution` loop didn't use a particular session, its id will be `nil`. Given that the only circumstance the id will be `nil` is if we don't need it anyways, I think we can just conditionally add the sessions with non-nil id's.

Reply:
Ok will do.